StripeIdentity/StripeIdentity/Source/NativeComponents/Views/PhoneOtpView.swift

```diff
@@ -0,0 +1,162 @@
+    func didInputFullOtp(newOtp: String)
+
+    func viewStateDidUpdate()
+}
+
+class PhoneOtpView: UIView {
+
+    enum ViewModel: Equatable {
+        case InputtingOTP // When user is inputting OTP
+        case SubmittingOTP(String) // When OTP is being submitted through Post VerificationPageData
+        case ErrorOTP // When wrong OTP is submitted
+        case RequestingOTP // When GenerateOTP is outstanding
+        case RequestingCannotVerify // When CannotVerify is outstanding
+    }
+
+    weak var delegate: PhoneOtpViewDelegate?
+
+    private let otpTextField: OneTimeCodeTextField
+    private let otpBodyLabel = {
+        let label = UILabel()
+        label.font = IdentityUI.instructionsFont
+        label.numberOfLines = 0
+        return label
+    }()
+    private let otpErrorLabel = {
+        let label = UILabel()
+        label.font = IdentityUI.instructionsFont
+        label.textColor = IdentityUI.identityElementsUITheme.colors.danger
+        label.numberOfLines = 1
+        return label
+    }()
+
+    private let otpErrorField = {
+        let stackView = UIStackView()
+        stackView.axis = .horizontal
+        stackView.alignment = .fill
+        stackView.distribution = .fill
+        stackView.spacing = 8
+        return stackView
+    }()
+
+    private let submittingIndicator = ActivityIndicator()
+
+    private let stackView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.axis = .vertical
+        stackView.alignment = .center
+        stackView.spacing = 16
+        return stackView
+    }()
+
+    var viewModel: ViewModel = .InputtingOTP
+
+    init(
+        otpLength: Int,
+        body: String,
+        errorString: String
+    ) {
+        otpTextField = OneTimeCodeTextField(
+            numberOfDigits: otpLength,
+            theme: IdentityUI.identityElementsUITheme
+        )
+        otpBodyLabel.text = body
+        otpErrorLabel.text = errorString
+
+        super.init(frame: .zero)
+        otpTextField.addTarget(self, action: #selector(otpTextFieldDidChange), for: .valueChanged)
+
+        let warningIconView = UIImageView()
+        warningIconView.contentMode = .scaleToFill
+        warningIconView.image = Image.iconWarning2.makeImage().withTintColor(.systemRed)
+
+        otpErrorField.addArrangedSubview(warningIconView)
+        otpErrorField.addArrangedSubview(otpErrorLabel)
+        stackView.addArrangedSubview(otpBodyLabel)
+        stackView.addArrangedSubview(otpTextField)
+        stackView.addArrangedSubview(otpErrorField)
+        stackView.addArrangedSubview(submittingIndicator)
+
+        NSLayoutConstraint.activate([
+            warningIconView.widthAnchor.constraint(equalToConstant: 16),
+            warningIconView.heightAnchor.constraint(equalToConstant: 16),
+            otpTextField.leadingAnchor.constraint(equalTo: stackView.leadingAnchor),
+            otpTextField.trailingAnchor.constraint(equalTo: stackView.trailingAnchor),
```

Comment:
double-checking: this is a little unexpected for me where `otpTextField` is added _within_ `UIStackView` so usually UIStackView would manage where its placed

what's the reasoning?

Reply:
@ccen-stripe 

yea sometimes layout like this gets confusing and easiest is to try a bunch of things

another way maybe is to put the error view in another stack view (that's centered)

but if this works go for it! I'm just not 100% sure on whether this could cause some layout issues (...because UIStackView is trying to do its own layout, yet we add additional constraints on top, which maybe conflict); worth also checking different iOS versions just-in-case

Reply:
`otpErrorField` is already a horizontal stackView 😄 , there might be some config to make it work as expected, but will stick to this one as it's tested thoroughly